cli/pack/common.go

```diff
@@ -0,0 +1,534 @@
+func findRocks(root string) (string, error) {
+	pattern := "*.rockspec"
+	res := ""
+	err := filepath.Walk(root, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		if info.IsDir() {
+			return nil
+		}
+		if matched, err := filepath.Match(pattern, filepath.Base(path)); err != nil {
+			return err
+		} else if matched {
+			info, err := os.Stat(path)
+			if err != nil {
+				return err
+			}
+			if info.IsDir() {
+				return nil
+			}
+			res = path
+		}
+		return nil
+	})
+	if err != nil {
+		return "", err
+	}
+	if res == "" {
+		return "", fmt.Errorf("rockspec not found")
+	}
+	return res, nil
+}
+
+func checkVersionFromGit(basePath string) (string, error) {
+	if basePath == "" {
+		return "", fmt.Errorf("passed zero path")
+	}
+	startPath, _ := os.Getwd()
+	defer func() {
+		_ = os.Chdir(startPath)
+	}()
+	err := os.Chdir(basePath)
+	if err != nil {
+		return "", err
+	}
+
+	cmd := exec.Command("git", "describe", "--tags", "--long")
+	var out bytes.Buffer
+	cmd.Stdout = &out
+	err = cmd.Run()
+	if err != nil {
+		return "", fmt.Errorf("no git version found")
+	}
+
+	res, err := normalizeGitVersion(out.String())
+	if err != nil {
+		return "", err
+	}
+
+	return res, nil
+}
+
+// normalizeGitVersion returns a version string filtered by regular expressions.
+func normalizeGitVersion(versionStr string) (string, error) {
+	expr1 := "\\d+\\.\\d+\\.\\d+"
+	expr2 := "\\-\\d+\\-"
+
+	regex1, err := regexp.Compile(expr1)
+	if err != nil {
+		return "", fmt.Errorf("regexp compile failed")
+	}
+	regex2, err := regexp.Compile(expr2)
+	if err != nil {
+		return "", fmt.Errorf("regexp compile failed")
+	}
+
+	majMinPatch := regex1.FindString(versionStr)
+	count := regex2.FindString(versionStr)
+	res := majMinPatch + "." + count[1:len(count)-1]
+
+	return res, nil
+}
+
+// collectAppList collects all the supposed applications from the passed directory,
+// considering the passed slice of exclude regexp items.
+func collectAppList(baseDir string, excludeList []*regexp.Regexp) ([]string, error) {
+	dirEnrties, err := os.ReadDir(baseDir)
+	if err != nil {
+		return nil, err
+	}
+
+	apps := make([]string, 0)
+	for _, entry := range dirEnrties {
+		if isApp(entry, excludeList) {
+			app := appNameFromEntry(entry)
+			if app != "" {
+				apps = append(apps, app)
+			}
+		}
+	}
+
+	if err != nil {
+		return nil, err
+	}
+	return apps, nil
+}
+
+// appNameFromEntry returns a normalized application name.
+// If the application is a lua file, the name of file will be returned
+// without its extension.
+func appNameFromEntry(entry os.DirEntry) string {
+	if filepath.Ext(entry.Name()) == ".lua" {
+		return entry.Name()[:len(entry.Name())-len(".lua")]
+	}
+	if entry.IsDir() || entry.Type() == os.ModeSymlink {
+		return entry.Name()
+	}
+	return ""
+}
+
+// isApp detects if the passed entry is an application.
+func isApp(entry os.DirEntry, excludeList []*regexp.Regexp) bool {
```

Comment:
Can we make this API more common? DirEntry argument makes a user to invoke ReadDir kind of functions to work with it. This is inconvenient. I'd like to have an API (in new application/app or existing util package) that allows me to test of the specified path is a Tarantool application.
Yes, we still can have a function, which accepts DirEntry in our pack code. But it must use this common function, which accepts a path to check.

Reply:
It is comfortable to use DirEntry, cause it is an interface, so easy to mock and this is an internal function, it has not much users, just me and maybe someone, who will fix it later.